src/server.rs

```diff
@@ -227,855 +725,233 @@ impl ServerStats {
-                        self.stats.cache_read_hit_duration += duration;
-                    },
-                    CompileResult::CacheMiss(miss_type, duration, future) => {
-                        match miss_type {
-                            MissType::Normal => self.stats.cache_misses += 1,
-                            MissType::CacheReadError => self.stats.cache_read_errors += 1,
-                            MissType::ForcedRecache => {
-                                self.stats.cache_misses += 1;
-                                self.stats.forced_recaches += 1;
-                            }
-                        }
-                        self.stats.cache_read_miss_duration += duration;
-                        self.await_cache_write(&mut event_loop, future)
-                    }
-                    CompileResult::NotCacheable => {
-                        self.stats.cache_misses += 1;
-                        self.stats.non_cacheable_compilations += 1;
-                    }
-                    CompileResult::CompileFailed => self.stats.compile_fails += 1,
-                };
-                let Output { status, stdout, stderr } = out;
-                status.code()
-                    .map_or_else(
-                        || trace!("CompileFinished missing retcode"),
-                        |s| { trace!("CompileFinished retcode: {}", s); finish.set_retcode(s) });
-                //TODO: sort out getting signal return on Unix
-                finish.set_stdout(stdout);
-                finish.set_stderr(stderr);
-            }
-            None => {
-                self.stats.cache_errors += 1;
-                //TODO: figure out a better way to communicate this?
-                finish.set_retcode(-2);
-            }
-        };
-        res.set_compile_finished(finish);
-        match self.conns[token].send(res, event_loop) {
-            Ok(_) => {}
-            Err(_) => {
-                error!("Failed to send response");
-            }
-        };
-    }
-
-
-    /// Check that `compiler` is `Some` and can handle `cmd`.
-    ///
-    /// If `cmd` is `Some` and does not contain unsupported options
-    /// (see `compiler_commandline_ok`), send the client a `CompileStarted`
-    /// message and begin compilation on a background task. Otherwise,
-    /// send the client an `UnhandledCompile` message.
-    fn check_compiler(&mut self, compiler: Option<Compiler>, cmd: Vec<String>, cwd: String, token: Token, event_loop: &mut EventLoop<SccacheServer<C>>) {
-        match compiler {
-            None => {
-                debug!("check_compiler: Unsupported compiler");
-                self.stats.requests_unsupported_compiler += 1;
-                self.send_unhandled_compile(token, event_loop);
+        match parse_length_delimited_from_bytes::<Request>(buf.as_slice()) {
+            Ok(req) => {
+                let size = req.write_length_delimited_to_bytes().unwrap().len();
```

Comment:
Yeah I tried using that but unfortunately it doesn't account for the header saying how big the request was, so it ended up causing an infinite loop with the codec :(